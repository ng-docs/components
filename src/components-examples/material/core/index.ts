import {NgModule} from '@angular/core';
import {FormsModule} from '@angular/forms';
import {MatButtonModule} from '@angular/material/button';
import {MatCheckboxModule} from '@angular/material/checkbox';
import {MatRippleModule} from '@angular/material/core';
import {MatInputModule} from '@angular/material/input';
import {ElevationOverviewExample} from './elevation-overview/elevation-overview-example';
import {RippleOverviewExample} from './ripple-overview/ripple-overview-example';

export {ElevationOverviewExample, RippleOverviewExample};

const EXAMPLES = [ElevationOverviewExample, RippleOverviewExample];

@NgModule({
  imports: [MatButtonModule, MatCheckboxModule, MatInputModule, MatRippleModule, FormsModule],
  declarations: EXAMPLES,
  exports: EXAMPLES,
})
export class CoreExamplesModule {}
