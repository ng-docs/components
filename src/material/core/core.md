The `@angular/material/core` package contains common components and services used by multiple other
components in the library. See the API tab for a listing of components and classes available.
