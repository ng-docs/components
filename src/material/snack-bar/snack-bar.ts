/**
 * @license
 * Copyright Google LLC All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */

import {LiveAnnouncer} from '@angular/cdk/a11y';
import {BreakpointObserver, Breakpoints} from '@angular/cdk/layout';
import {Overlay, OverlayConfig, OverlayRef} from '@angular/cdk/overlay';
import {ComponentPortal, ComponentType, TemplatePortal} from '@angular/cdk/portal';
import {
  ComponentRef,
  EmbeddedViewRef,
  Inject,
  Injectable,
  InjectionToken,
  Injector,
  Optional,
  SkipSelf,
  TemplateRef,
  OnDestroy,
  Type,
} from '@angular/core';
import {takeUntil} from 'rxjs/operators';
import {TextOnlySnackBar, SimpleSnackBar} from './simple-snack-bar';
import {MAT_SNACK_BAR_DATA, MatSnackBarConfig} from './snack-bar-config';
import {MatSnackBarContainer, _MatSnackBarContainerBase} from './snack-bar-container';
import {MatSnackBarModule} from './snack-bar-module';
import {MatSnackBarRef} from './snack-bar-ref';

/**
 * Injection token that can be used to specify default snack bar.
 *
 * 这个注入令牌可以用来指定快餐栏的默认配置。
 *
 */
export const MAT_SNACK_BAR_DEFAULT_OPTIONS = new InjectionToken<MatSnackBarConfig>(
  'mat-snack-bar-default-options',
  {
    providedIn: 'root',
    factory: MAT_SNACK_BAR_DEFAULT_OPTIONS_FACTORY,
  },
);

/** @docs-private */
export function MAT_SNACK_BAR_DEFAULT_OPTIONS_FACTORY(): MatSnackBarConfig {
  return new MatSnackBarConfig();
}

@Injectable()
export abstract class _MatSnackBarBase implements OnDestroy {
  /**
   * Reference to the current snack bar in the view *at this level* (in the Angular injector tree).
   * If there is a parent snack-bar service, all operations should delegate to that parent
   * via `_openedSnackBarRef`.
   */
  private _snackBarRefAtThisLevel: MatSnackBarRef<any> | null = null;

  /**
   * The component that should be rendered as the snack bar's simple component.
   *
   * 该组件应该被渲染成快餐栏的简单组件。
   *
   */
  protected abstract simpleSnackBarComponent: Type<TextOnlySnackBar>;

  /**
   * The container component that attaches the provided template or component.
   *
   * 附加在所提供的模板或组件上的容器组件。
   *
   */
  protected abstract snackBarContainerComponent: Type<_MatSnackBarContainerBase>;

  /**
   * The CSS class to apply for handset mode.
   *
   * 适用于手机模式的 CSS 类。
   *
   */
  protected abstract handsetCssClass: string;

  /** Reference to the currently opened snackbar at *any* level. */
  get _openedSnackBarRef(): MatSnackBarRef<any> | null {
    const parent = this._parentSnackBar;
    return parent ? parent._openedSnackBarRef : this._snackBarRefAtThisLevel;
  }

  set _openedSnackBarRef(value: MatSnackBarRef<any> | null) {
    if (this._parentSnackBar) {
      this._parentSnackBar._openedSnackBarRef = value;
    } else {
      this._snackBarRefAtThisLevel = value;
    }
  }

  constructor(
    private _overlay: Overlay,
    private _live: LiveAnnouncer,
    private _injector: Injector,
    private _breakpointObserver: BreakpointObserver,
    @Optional() @SkipSelf() private _parentSnackBar: _MatSnackBarBase,
    @Inject(MAT_SNACK_BAR_DEFAULT_OPTIONS) private _defaultConfig: MatSnackBarConfig,
  ) {}

  /**
   * Creates and dispatches a snack bar with a custom component for the content, removing any
   * currently opened snack bars.
   *
   * 为内容创建并派发带有自定义组件的快餐栏，删除任何目前打开着的快餐栏。
   *
   * @param component Component to be instantiated.
   *
   * 要实例化的组件。
   *
   * @param config Extra configuration for the snack bar.
   *
   * 快餐栏的额外配置。
   *
   */
  openFromComponent<T, D = any>(
    component: ComponentType<T>,
    config?: MatSnackBarConfig<D>,
  ): MatSnackBarRef<T> {
    return this._attach(component, config) as MatSnackBarRef<T>;
  }

  /**
   * Creates and dispatches a snack bar with a custom template for the content, removing any
   * currently opened snack bars.
   *
   * 使用该内容的自定义模板创建并派发一个快餐栏，删除任何目前打开着的快餐栏。
   *
   * @param template Template to be instantiated.
   *
   * 要实例化的模板。
   *
   * @param config Extra configuration for the snack bar.
   *
   * 快餐栏的额外配置。
   *
   */
  openFromTemplate(
    template: TemplateRef<any>,
    config?: MatSnackBarConfig,
  ): MatSnackBarRef<EmbeddedViewRef<any>> {
    return this._attach(template, config);
  }

  /**
   * Opens a snackbar with a message and an optional action.
   *
   * 带有信息和可选操作的快餐栏。
   *
   * @param message The message to show in the snackbar.
   *
   * 要显示在快餐栏里的消息。
   *
   * @param action The label for the snackbar action.
   *
   * 快餐栏的动作标签。
   *
   * @param config Additional configuration options for the snackbar.
   *
   * 快餐栏的其它配置选项。
   *
   */
  open(
    message: string,
    action: string = '',
    config?: MatSnackBarConfig,
  ): MatSnackBarRef<TextOnlySnackBar> {
    const _config = {...this._defaultConfig, ...config};

    // Since the user doesn't have access to the component, we can
    // override the data to pass in our own message and action.
    _config.data = {message, action};

    // Since the snack bar has `role="alert"`, we don't
    // want to announce the same message twice.
    if (_config.announcementMessage === message) {
      _config.announcementMessage = undefined;
    }

    return this.openFromComponent(this.simpleSnackBarComponent, _config);
  }

  /**
   * Dismisses the currently-visible snack bar.
   *
   * 关闭目前可见的快餐栏。
   *
   */
  dismiss(): void {
    if (this._openedSnackBarRef) {
      this._openedSnackBarRef.dismiss();
    }
  }

  ngOnDestroy() {
    // Only dismiss the snack bar at the current level on destroy.
    if (this._snackBarRefAtThisLevel) {
      this._snackBarRefAtThisLevel.dismiss();
    }
  }

  /**
   * Attaches the snack bar container component to the overlay.
   */
  private _attachSnackBarContainer(
    overlayRef: OverlayRef,
    config: MatSnackBarConfig,
  ): _MatSnackBarContainerBase {
    const userInjector = config && config.viewContainerRef && config.viewContainerRef.injector;
    const injector = Injector.create({
      parent: userInjector || this._injector,
      providers: [{provide: MatSnackBarConfig, useValue: config}],
    });

    const containerPortal = new ComponentPortal(
      this.snackBarContainerComponent,
      config.viewContainerRef,
      injector,
    );
    const containerRef: ComponentRef<_MatSnackBarContainerBase> =
      overlayRef.attach(containerPortal);
    containerRef.instance.snackBarConfig = config;
    return containerRef.instance;
  }

  /**
   * Places a new component or a template as the content of the snack bar container.
   */
  private _attach<T>(
    content: ComponentType<T> | TemplateRef<T>,
    userConfig?: MatSnackBarConfig,
  ): MatSnackBarRef<T | EmbeddedViewRef<any>> {
    const config = {...new MatSnackBarConfig(), ...this._defaultConfig, ...userConfig};
    const overlayRef = this._createOverlay(config);
    const container = this._attachSnackBarContainer(overlayRef, config);
    const snackBarRef = new MatSnackBarRef<T | EmbeddedViewRef<any>>(container, overlayRef);

    if (content instanceof TemplateRef) {
      const portal = new TemplatePortal(content, null!, {
        $implicit: config.data,
        snackBarRef,
      } as any);

      snackBarRef.instance = container.attachTemplatePortal(portal);
    } else {
      const injector = this._createInjector(config, snackBarRef);
      const portal = new ComponentPortal(content, undefined, injector);
      const contentRef = container.attachComponentPortal<T>(portal);

      // We can't pass this via the injector, because the injector is created earlier.
      snackBarRef.instance = contentRef.instance;
    }

    // Subscribe to the breakpoint observer and attach the mat-snack-bar-handset class as
    // appropriate. This class is applied to the overlay element because the overlay must expand to
    // fill the width of the screen for full width snackbars.
    this._breakpointObserver
      .observe(Breakpoints.HandsetPortrait)
      .pipe(takeUntil(overlayRef.detachments()))
      .subscribe(state => {
        overlayRef.overlayElement.classList.toggle(this.handsetCssClass, state.matches);
      });

    if (config.announcementMessage) {
      // Wait until the snack bar contents have been announced then deliver this message.
      container._onAnnounce.subscribe(() => {
        this._live.announce(config.announcementMessage!, config.politeness);
      });
    }

    this._animateSnackBar(snackBarRef, config);
    this._openedSnackBarRef = snackBarRef;
    return this._openedSnackBarRef;
  }

  /** Animates the old snack bar out and the new one in. */
  private _animateSnackBar(snackBarRef: MatSnackBarRef<any>, config: MatSnackBarConfig) {
    // When the snackbar is dismissed, clear the reference to it.
    snackBarRef.afterDismissed().subscribe(() => {
      // Clear the snackbar ref if it hasn't already been replaced by a newer snackbar.
      if (this._openedSnackBarRef == snackBarRef) {
        this._openedSnackBarRef = null;
      }

      if (config.announcementMessage) {
        this._live.clear();
      }
    });

    if (this._openedSnackBarRef) {
      // If a snack bar is already in view, dismiss it and enter the
      // new snack bar after exit animation is complete.
      this._openedSnackBarRef.afterDismissed().subscribe(() => {
        snackBarRef.containerInstance.enter();
      });
      this._openedSnackBarRef.dismiss();
    } else {
      // If no snack bar is in view, enter the new snack bar.
      snackBarRef.containerInstance.enter();
    }

    // If a dismiss timeout is provided, set up dismiss based on after the snackbar is opened.
    if (config.duration && config.duration > 0) {
      snackBarRef.afterOpened().subscribe(() => snackBarRef._dismissAfter(config.duration!));
    }
  }

  /**
   * Creates a new overlay and places it in the correct location.
   * @param config The user-specified snack bar config.
   */
  private _createOverlay(config: MatSnackBarConfig): OverlayRef {
    const overlayConfig = new OverlayConfig();
    overlayConfig.direction = config.direction;

    let positionStrategy = this._overlay.position().global();
    // Set horizontal position.
    const isRtl = config.direction === 'rtl';
    const isLeft =
      config.horizontalPosition === 'left' ||
      (config.horizontalPosition === 'start' && !isRtl) ||
      (config.horizontalPosition === 'end' && isRtl);
    const isRight = !isLeft && config.horizontalPosition !== 'center';
    if (isLeft) {
      positionStrategy.left('0');
    } else if (isRight) {
      positionStrategy.right('0');
    } else {
      positionStrategy.centerHorizontally();
    }
    // Set horizontal position.
    if (config.verticalPosition === 'top') {
      positionStrategy.top('0');
    } else {
      positionStrategy.bottom('0');
    }

    overlayConfig.positionStrategy = positionStrategy;
    return this._overlay.create(overlayConfig);
  }

  /**
   * Creates an injector to be used inside of a snack bar component.
   * @param config Config that was used to create the snack bar.
   * @param snackBarRef Reference to the snack bar.
   */
  private _createInjector<T>(config: MatSnackBarConfig, snackBarRef: MatSnackBarRef<T>): Injector {
    const userInjector = config && config.viewContainerRef && config.viewContainerRef.injector;

    return Injector.create({
      parent: userInjector || this._injector,
      providers: [
        {provide: MatSnackBarRef, useValue: snackBarRef},
        {provide: MAT_SNACK_BAR_DATA, useValue: config.data},
      ],
    });
  }
}

/**
 * Service to dispatch Material Design snack bar messages.
 *
 * 一个服务，用来派发 Material Design 快餐栏的消息。
 *
 */
@Injectable({providedIn: MatSnackBarModule})
export class MatSnackBar extends _MatSnackBarBase {
  protected simpleSnackBarComponent = SimpleSnackBar;
  protected snackBarContainerComponent = MatSnackBarContainer;
  protected handsetCssClass = 'mat-snack-bar-handset';

  constructor(
    overlay: Overlay,
    live: LiveAnnouncer,
    injector: Injector,
    breakpointObserver: BreakpointObserver,
    @Optional() @SkipSelf() parentSnackBar: MatSnackBar,
    @Inject(MAT_SNACK_BAR_DEFAULT_OPTIONS) defaultConfig: MatSnackBarConfig,
  ) {
    super(overlay, live, injector, breakpointObserver, parentSnackBar, defaultConfig);
  }
}
