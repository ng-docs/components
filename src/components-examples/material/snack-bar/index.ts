import {NgModule} from '@angular/core';
import {FormsModule} from '@angular/forms';
import {MatButtonModule} from '@angular/material/button';
import {MatInputModule} from '@angular/material/input';
import {MatSelectModule} from '@angular/material/select';
import {MatSnackBarModule} from '@angular/material/snack-bar';
import {
  PizzaPartyComponent,
  SnackBarComponentExample,
} from './snack-bar-component/snack-bar-component-example';
import {
  PizzaPartyAnnotatedComponent,
  SnackBarAnnotatedComponentExample,
} from './snack-bar-annotated-component/snack-bar-annotated-component-example';
import {SnackBarOverviewExample} from './snack-bar-overview/snack-bar-overview-example';
import {SnackBarPositionExample} from './snack-bar-position/snack-bar-position-example';
import {SnackBarHarnessExample} from './snack-bar-harness/snack-bar-harness-example';

export {
  SnackBarAnnotatedComponentExample,
  SnackBarComponentExample,
  SnackBarHarnessExample,
  SnackBarOverviewExample,
  SnackBarPositionExample,
  PizzaPartyComponent,
  PizzaPartyAnnotatedComponent,
};

const EXAMPLES = [
  SnackBarComponentExample,
  SnackBarHarnessExample,
  SnackBarOverviewExample,
  SnackBarPositionExample,
  SnackBarAnnotatedComponentExample,
];

@NgModule({
  imports: [FormsModule, MatButtonModule, MatInputModule, MatSelectModule, MatSnackBarModule],
  declarations: [...EXAMPLES, PizzaPartyComponent, PizzaPartyAnnotatedComponent],
  exports: EXAMPLES,
})
export class SnackBarExamplesModule {}
