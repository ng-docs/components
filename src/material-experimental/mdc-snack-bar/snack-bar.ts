/**
 * @license
 * Copyright Google LLC All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */

import {Injectable} from '@angular/core';
import {MatSnackBar as BaseMatSnackBar} from '@angular/material/snack-bar';
import {MatSnackBarModule} from './module';
import {MatSimpleSnackBar} from './simple-snack-bar';
import {MatSnackBarContainer} from './snack-bar-container';

/**
 * Service to dispatch Material Design snack bar messages.
 */
@Injectable({providedIn: MatSnackBarModule})
export class MatSnackBar extends BaseMatSnackBar {
  protected override simpleSnackBarComponent = MatSimpleSnackBar;
  protected override snackBarContainerComponent = MatSnackBarContainer;
  protected override handsetCssClass = 'mat-mdc-snack-bar-handset';
}
