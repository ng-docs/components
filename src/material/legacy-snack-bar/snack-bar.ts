/**
 * @license
 * Copyright Google LLC All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */

import {LiveAnnouncer} from '@angular/cdk/a11y';
import {BreakpointObserver} from '@angular/cdk/layout';
import {Overlay} from '@angular/cdk/overlay';
import {Inject, Injectable, Injector, Optional, SkipSelf} from '@angular/core';
import {LegacySimpleSnackBar} from './simple-snack-bar';
import {
  _MatSnackBarBase,
  MAT_SNACK_BAR_DEFAULT_OPTIONS,
  MatSnackBarConfig,
} from '@angular/material/snack-bar';
import {MatLegacySnackBarContainer} from './snack-bar-container';
import {MatLegacySnackBarModule} from './snack-bar-module';

/**
 * Service to dispatch Material Design snack bar messages.
 * @deprecated Use `MatSnackBar` from `@angular/material/snack-bar` instead. See https://material.angular.io/guide/mdc-migration for information about migrating.
 * @breaking-change 17.0.0
 */
@Injectable({providedIn: MatLegacySnackBarModule})
export class MatLegacySnackBar extends _MatSnackBarBase {
  protected simpleSnackBarComponent = LegacySimpleSnackBar;
  protected snackBarContainerComponent = MatLegacySnackBarContainer;
  protected handsetCssClass = 'mat-snack-bar-handset';

  constructor(
    overlay: Overlay,
    live: LiveAnnouncer,
    injector: Injector,
    breakpointObserver: BreakpointObserver,
    @Optional() @SkipSelf() parentSnackBar: MatLegacySnackBar,
    @Inject(MAT_SNACK_BAR_DEFAULT_OPTIONS) defaultConfig: MatSnackBarConfig,
  ) {
    super(overlay, live, injector, breakpointObserver, parentSnackBar, defaultConfig);
  }
}
