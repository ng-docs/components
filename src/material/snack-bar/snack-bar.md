`MatSnackBar` is a service for displaying snack-bar notifications.

`MatSnackBar` 是一个用来显示快餐栏通知的服务。

<!-- example(snack-bar-overview) -->

### Opening a snackbar

### 打开快餐栏

A snackbar can contain either a string message or a given component.

快餐栏可以包含一个字符串消息或指定的组件。

```ts
// Simple message.
let snackBarRef = snackBar.open('Message archived');

// Simple message with an action.
let snackBarRef = snackBar.open('Message archived', 'Undo');

// Load the given component into the snackbar.
let snackBarRef = snackbar.openFromComponent(MessageArchivedComponent);
```

In either case, a `MatSnackBarRef` is returned. This can be used to dismiss the snackbar or to
receive notification of when the snackbar is dismissed. For simple messages with an action, the
`MatSnackBarRef` exposes an observable for when the action is triggered.
If you want to close a custom snackbar that was opened via `openFromComponent`, from within the
component itself, you can inject the `MatSnackBarRef`.

无论哪种形式，都会返回一个 `MatSnackBarRef`。它可以用来关闭快餐栏或在快餐栏关闭时接收通知。
对于只有一个操作的简单消息，当该动作被触发时，`MatSnackBarRef` 会暴露出一个 `Observable`。
如果你要关闭一个用 `openFromComponent` 打开的自定义快餐栏，可以在该组件中注入一个 `MatSnackBarRef`。

```ts
snackBarRef.afterDismissed().subscribe(() => {
  console.log('The snackbar was dismissed');
});

snackBarRef.onAction().subscribe(() => {
  console.log('The snackbar action was triggered!');
});

snackBarRef.dismiss();
```

### Dismissal

### 关闭

A snackbar can be dismissed manually by calling the `dismiss` method on the `MatSnackBarRef`
returned from the call to `open`.

可以调用由 `open` 调用返回的 `MatSnackBarRef` 中的 `dismiss` 方法来手动关闭快餐栏。

Only one snackbar can ever be opened at one time. If a new snackbar is opened while a previous
message is still showing, the older message will be automatically dismissed.

同一时刻只能打开一个快餐栏。如果在显示前一个消息时打开一个新的快餐栏，老的消息就会自动关闭。

A snackbar can also be given a duration via the optional configuration object:

快餐栏还可以通过一个可选的配置对象来指定持续时间：

```ts
snackbar.open('Message archived', 'Undo', {
  duration: 3000
});
```

### Sharing data with a custom snackbar

### 与自定义快餐栏共享数据

You can share data with the custom snackbar, that you opened via the `openFromComponent` method,
by passing it through the `data` property.

你可以传入 `data` 属性，来与 `openFromComponent` 打开的自定义快餐栏共享数据。

```ts
snackbar.openFromComponent(MessageArchivedComponent, {
  data: 'some data'
});
```

To access the data in your component, you have to use the `MAT_SNACK_BAR_DATA` injection token:

要在组件中访问该数据，可以使用依赖注入令牌 `MAT_SNACK_BAR_DATA`：

```ts
import {Component, Inject} from '@angular/core';
import {MAT_SNACK_BAR_DATA} from '@angular/material/snack-bar';

@Component({
  selector: 'your-snackbar',
  template: 'passed in {{ data }}',
})
export class MessageArchivedComponent {
  constructor(@Inject(MAT_SNACK_BAR_DATA) public data: string) { }
}
```

### Setting the global configuration defaults

### 设置全局配置的默认值

If you want to override the default snack bar options, you can do so using the
`MAT_SNACK_BAR_DEFAULT_OPTIONS` injection token.

如果你要覆盖快餐栏的默认选项，可以使用 `MAT_SNACK_BAR_DEFAULT_OPTIONS` 令牌。

```ts
@NgModule({
  providers: [
    {provide: MAT_SNACK_BAR_DEFAULT_OPTIONS, useValue: {duration: 2500}}
  ]
})
```

### Accessibility

### 无障碍性

`MatSnackBar` announces messages via an `aria-live` region. While announcements use the `polite`
setting by default, you can customize this by setting the `politeness` property of
`MatSnackBarConfig`.

`MatSnackBar` 通过 `aria-live` 区域来播报消息。虽然默认情况下公告会使用 `polite` 设置，但你也可以通过设置 `MatSnackBarConfig` 的 `politeness` 属性来自定义它。

`MatSnackBar` does not move focus to the snackbar element. Moving focus like this would disrupt
users in the middle of a workflow. For any action offered in the snackbar, your application should
provide an alternative way to perform the action. Alternative interactions are typically keyboard
shortcuts or menu options. You should dismiss the snackbar once the user performs its corresponding
action. A snackbar can contain a single action with an additional optional "dismiss" or "cancel"
action.

`MatSnackBar` 不会将焦点移动到快餐栏元素。像这样移动焦点会在工作流程中扰乱用户。对于快餐栏中提供的任何操作，你的应用程序都应该提供执行该操作的替代方法。替代交互通常是键盘快捷键或菜单选项。一旦用户执行相应的操作，你应该关闭快餐栏。快餐栏可以包含一个单独的动作和一个额外的可选“关闭”或“取消”动作。

Avoid setting a `duration` for snackbars that have an action available, as screen reader users may
want to navigate to the snackbar element to activate the action. If the user has manually moved
their focus within the snackbar, you should return focus somewhere that makes sense in the context
of the user's workflow.

尽量不要为具有可用操作的快餐栏设置 `duration` ，因为屏幕阅读器用户可能希望导航到快餐栏上的元素以激活该操作。如果用户在快餐栏中手动移动了他们的焦点，你应该将焦点返回到用户工作流程上下文中有意义的某个地方。
