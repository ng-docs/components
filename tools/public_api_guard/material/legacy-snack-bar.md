## API Report File for "components-srcs"

> Do not edit this file. It is a report generated by [API Extractor](https://api-extractor.com/).

```ts

import { BreakpointObserver } from '@angular/cdk/layout';
import * as i0 from '@angular/core';
import * as i3 from '@angular/cdk/overlay';
import * as i4 from '@angular/cdk/portal';
import * as i5 from '@angular/common';
import * as i6 from '@angular/material/legacy-button';
import * as i7 from '@angular/material/core';
import { Injector } from '@angular/core';
import { TextOnlySnackBar as LegacyTextOnlySnackBar } from '@angular/material/snack-bar';
import { LiveAnnouncer } from '@angular/cdk/a11y';
import { MAT_SNACK_BAR_DATA as MAT_LEGACY_SNACK_BAR_DATA } from '@angular/material/snack-bar';
import { MAT_SNACK_BAR_DEFAULT_OPTIONS as MAT_LEGACY_SNACK_BAR_DEFAULT_OPTIONS } from '@angular/material/snack-bar';
import { MAT_SNACK_BAR_DEFAULT_OPTIONS_FACTORY as MAT_LEGACY_SNACK_BAR_DEFAULT_OPTIONS_FACTORY } from '@angular/material/snack-bar';
import { matSnackBarAnimations as matLegacySnackBarAnimations } from '@angular/material/snack-bar';
import { _MatSnackBarBase as _MatLegacySnackBarBase } from '@angular/material/snack-bar';
import { MatSnackBarConfig as MatLegacySnackBarConfig } from '@angular/material/snack-bar';
import { _MatSnackBarContainerBase as _MatLegacySnackBarContainerBase } from '@angular/material/snack-bar';
import { MatSnackBarDismiss as MatLegacySnackBarDismiss } from '@angular/material/snack-bar';
import { MatSnackBarHorizontalPosition as MatLegacySnackBarHorizontalPosition } from '@angular/material/snack-bar';
import { MatSnackBarRef as MatLegacySnackBarRef } from '@angular/material/snack-bar';
import { MatSnackBarVerticalPosition as MatLegacySnackBarVerticalPosition } from '@angular/material/snack-bar';
import { Overlay } from '@angular/cdk/overlay';

// @public @deprecated
export class LegacySimpleSnackBar implements LegacyTextOnlySnackBar {
    constructor(snackBarRef: MatLegacySnackBarRef<LegacySimpleSnackBar>, data: any);
    action(): void;
    data: {
        message: string;
        action: string;
    };
    get hasAction(): boolean;
    // (undocumented)
    snackBarRef: MatLegacySnackBarRef<LegacySimpleSnackBar>;
    // (undocumented)
    static ɵcmp: i0.ɵɵComponentDeclaration<LegacySimpleSnackBar, "simple-snack-bar", never, {}, {}, never, never, false, never>;
    // (undocumented)
    static ɵfac: i0.ɵɵFactoryDeclaration<LegacySimpleSnackBar, never>;
}

export { LegacyTextOnlySnackBar }

export { MAT_LEGACY_SNACK_BAR_DATA }

export { MAT_LEGACY_SNACK_BAR_DEFAULT_OPTIONS }

export { MAT_LEGACY_SNACK_BAR_DEFAULT_OPTIONS_FACTORY }

// @public @deprecated
export class MatLegacySnackBar extends _MatLegacySnackBarBase {
    constructor(overlay: Overlay, live: LiveAnnouncer, injector: Injector, breakpointObserver: BreakpointObserver, parentSnackBar: MatLegacySnackBar, defaultConfig: MatLegacySnackBarConfig);
    // (undocumented)
    protected handsetCssClass: string;
    // (undocumented)
    protected simpleSnackBarComponent: typeof LegacySimpleSnackBar;
    // (undocumented)
    protected snackBarContainerComponent: typeof MatLegacySnackBarContainer;
    // (undocumented)
    static ɵfac: i0.ɵɵFactoryDeclaration<MatLegacySnackBar, [null, null, null, null, { optional: true; skipSelf: true; }, null]>;
    // (undocumented)
    static ɵprov: i0.ɵɵInjectableDeclaration<MatLegacySnackBar>;
}

export { matLegacySnackBarAnimations }

export { _MatLegacySnackBarBase }

export { MatLegacySnackBarConfig }

// @public @deprecated
export class MatLegacySnackBarContainer extends _MatLegacySnackBarContainerBase {
    // (undocumented)
    protected _afterPortalAttached(): void;
    // (undocumented)
    static ɵcmp: i0.ɵɵComponentDeclaration<MatLegacySnackBarContainer, "snack-bar-container", never, {}, {}, never, never, false, never>;
    // (undocumented)
    static ɵfac: i0.ɵɵFactoryDeclaration<MatLegacySnackBarContainer, never>;
}

export { _MatLegacySnackBarContainerBase }

export { MatLegacySnackBarDismiss }

export { MatLegacySnackBarHorizontalPosition }

// @public @deprecated (undocumented)
export class MatLegacySnackBarModule {
    // (undocumented)
    static ɵfac: i0.ɵɵFactoryDeclaration<MatLegacySnackBarModule, never>;
    // (undocumented)
    static ɵinj: i0.ɵɵInjectorDeclaration<MatLegacySnackBarModule>;
    // (undocumented)
    static ɵmod: i0.ɵɵNgModuleDeclaration<MatLegacySnackBarModule, [typeof i1.MatLegacySnackBarContainer, typeof i2.LegacySimpleSnackBar], [typeof i3.OverlayModule, typeof i4.PortalModule, typeof i5.CommonModule, typeof i6.MatLegacyButtonModule, typeof i7.MatCommonModule], [typeof i1.MatLegacySnackBarContainer, typeof i7.MatCommonModule]>;
}

export { MatLegacySnackBarRef }

export { MatLegacySnackBarVerticalPosition }

// (No @packageDocumentation comment for this package)

```
