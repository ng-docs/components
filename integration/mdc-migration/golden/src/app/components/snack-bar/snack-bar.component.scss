::ng-deep {
  .mat-mdc-snack-bar-container { background: lavender; }
  .mat-mdc-simple-snack-bar {
    color: darkblue;
  }
}
