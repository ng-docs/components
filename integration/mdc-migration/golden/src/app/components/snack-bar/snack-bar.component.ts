import {Component} from '@angular/core';
import {MatSnackBar} from '@angular/material/snack-bar';

@Component({
  selector: 'snack-bar-example',
  templateUrl: './snack-bar.component.html',
  styleUrls: ['./snack-bar.component.scss'],
})
export class SnackBarComponent {
  constructor(private _snackBar: MatSnackBar) {}

  openSnackBar(message: string) {
    this._snackBar.open(message, 'Dismiss');
  }
}
