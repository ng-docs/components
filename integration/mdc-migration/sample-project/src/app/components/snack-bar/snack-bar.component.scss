::ng-deep {
  .mat-snack-bar-container { background: lavender; }
  .mat-simple-snackbar {
    color: darkblue;
  }
}
