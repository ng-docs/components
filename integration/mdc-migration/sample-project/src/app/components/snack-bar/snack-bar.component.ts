import {Component} from '@angular/core';
import {MatLegacySnackBar as MatSnackBar} from '@angular/material/legacy-snack-bar';

@Component({
  selector: 'snack-bar-example',
  templateUrl: './snack-bar.component.html',
  styleUrls: ['./snack-bar.component.scss'],
})
export class SnackBarComponent {
  constructor(private _snackBar: MatSnackBar) {}

  openSnackBar(message: string) {
    this._snackBar.open(message, 'Dismiss');
  }
}
