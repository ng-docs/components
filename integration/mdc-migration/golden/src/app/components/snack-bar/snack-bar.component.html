<h2>Snackbar example</h2>
<mat-form-field appearance="fill">
  <mat-label>Message</mat-label>
  <input matInput value="Disco party!" #message>
</mat-form-field>
<button mat-stroked-button (click)="openSnackBar(message.value)">Show snack-bar</button>
